homeassistant/components/yeelight/__init__.py

```diff
@@ -139,73 +148,192 @@
     "active_mode",
 ]
 
+PLATFORMS = ["binary_sensor", "light"]
+
 
-def setup(hass, config):
+async def async_setup(hass: HomeAssistant, config: dict) -> bool:
     """Set up the Yeelight bulbs."""
     conf = config.get(DOMAIN, {})
-    yeelight_data = hass.data[DATA_YEELIGHT] = {}
+    hass.data[DOMAIN] = {
+        DATA_CUSTOM_EFFECTS: conf.get(CONF_CUSTOM_EFFECTS, {}),
+        DATA_CONFIG_ENTRIES: {},
+        DATA_SCAN_INTERVAL: conf.get(CONF_SCAN_INTERVAL, SCAN_INTERVAL),
+    }
 
-    def device_discovered(_, info):
-        _LOGGER.debug("Adding autodetected %s", info["hostname"])
+    # Import manually configured devices
+    for ipaddr, device_config in config.get(DOMAIN, {}).get(CONF_DEVICES, {}).items():
+        _LOGGER.debug("Importing configured %s", ipaddr)
+        entry_config = {
+            CONF_IP_ADDRESS: ipaddr,
+            **device_config,
+        }
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=entry_config,
+            ),
+        )
 
-        name = "yeelight_{}_{}".format(info["device_type"], info["properties"]["mac"])
+    return True
 
-        device_config = DEVICE_SCHEMA({CONF_NAME: name})
 
-        _setup_device(hass, config, info[CONF_HOST], device_config)
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Yeelight from a config entry."""
 
-    discovery.listen(hass, SERVICE_YEELIGHT, device_discovered)
+    async def _initialize(ipaddr: str) -> None:
+        device = await _async_setup_device(hass, ipaddr, entry.data)
+        hass.data[DOMAIN][DATA_CONFIG_ENTRIES][entry.entry_id] = device
+        for component in PLATFORMS:
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(entry, component)
+            )
 
-    def update(_):
-        for device in list(yeelight_data.values()):
-            device.update()
+    if entry.data[CONF_IP_ADDRESS]:
+        # manually added device
+        await _initialize(entry.data[CONF_IP_ADDRESS])
+    else:
+        # discovery
```

Comment:
A discovered entry should also have an ip address. That was what we decided, right? All config entries should have the same structure.

Reply:
But if the IP address is added to the config entry, the device is required to have a static IP. Currently with discovery, we can recognize devices with unique IDs.

Reply:
Ok.